### PR TITLE
Allow building with gcc on macOS

### PR DESCRIPTION
**Pull request by barracuda156**
@HofiOne Could you please check if this works correctly for you or something must be changed?

What I want to achieve is leave default behavior unchanged, but allow a manual configuration to work correctly.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,7 +18,19 @@ project(syslog-ng C)
...
+  set(CMAKE_CXX_STANDARD_REQUIRED OFF CACHE BOOL "C++ standard is a requirement")
+  set(CMAKE_CXX_EXTENSIONS OFF CACHE BOOL "C++ compiler specific extensions")
+  set(SYSLOG_NG_ENABLE_CPP ${ENABLE_CPP})
+endif()
+
 if(APPLE)
+  option(ENABLE_DARWIN_OSL "Enable OSLog" ON)
+endif()
+
+if(ENABLE_DARWIN_OSL)
```
Comment (HofiOne) [inline CMakeLists.txt:33]: and, use the new ENABLE_OBJC here instead

File: modules/darwinosl/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-if(APPLE)
+if(ENABLE_DARWIN_OSL)
```
Comment (HofiOne) [inline modules/darwinosl/CMakeLists.txt:1]: and, use the new ENABLE_OBJC here instead as well

File: CMakeLists.txt [hunk 3]
```diff
@@ -18,7 +18,19 @@ project(syslog-ng C)
...
+if (ENABLE_CPP)
+  enable_language(CXX)
+  set(CMAKE_CXX_STANDARD 14 CACHE STRING "C++ standard")
+  set(CMAKE_CXX_STANDARD_REQUIRED OFF CACHE BOOL "C++ standard is a requirement")
+  set(CMAKE_CXX_EXTENSIONS OFF CACHE BOOL "C++ compiler specific extensions")
+  set(SYSLOG_NG_ENABLE_CPP ${ENABLE_CPP})
+endif()
+
 if(APPLE)
+  option(ENABLE_DARWIN_OSL "Enable OSLog" ON)
```
Comment (HofiOne) [inline CMakeLists.txt:30]: i suggest adding a new variable as well here e.g.

ENABLE_OBJC

or s.g. like that, and set it based on the status of ENABLE_DARWIN_OSL (and later also based on any optional module-enabled settings that depend on objc)

option(ENABLE_OBJC "Enable OSLog" ENABLE_DARWIN_OSL)

File: modules/darwinosl/CMakeLists.txt [hunk 4]
```diff
@@ -26,11 +26,9 @@ if(APPLE)
     #include <OSLog/OSLog.h>
     int main(int, char**)
     { return 0; }
-  " HAVE_OSLLOG)
+  " HAVE_OSLOG)
```
Comment (HofiOne) [inline modules/darwinosl/CMakeLists.txt:29]: and, then also make this whole test guarded by the status of ENABLE_DARWIN_OSL

**Review by HofiOne — CHANGES_REQUESTED** [review]

**Comment by barracuda156** [comment]
@HofiOne I am not really sure if this is what you meant. Could you please check and correct what is needed to be corrected? Either just add a commit into the PR or tell me what to change. (To unbreak gcc build this version also works.)

File: modules/darwinosl/CMakeLists.txt [hunk 5]
```diff
@@ -1,4 +1,4 @@
-if(APPLE)
+if(ENABLE_OBJC)
```
Comment (HofiOne) [inline modules/darwinosl/CMakeLists.txt:1]: I think this must be

if (ENABLE_OBJC AND ENABLE_DARWIN_OSL)

**Review by HofiOne — CHANGES_REQUESTED** [review]

File: modules/darwinosl/CMakeLists.txt (hunk 5 above)
Comment (barracuda156) [reply modules/darwinosl/CMakeLists.txt:1]: But they are not independent. `ENABLE_DARWIN_OSL` cannot work without ObjC, and `ENABLE_OBJC` is set by `ENABLE_DARWIN_OSL` value. Do I miss something?

File: modules/darwinosl/CMakeLists.txt (hunk 4 above)
Comment (barracuda156) [reply modules/darwinosl/CMakeLists.txt:29]: Is there any point in enabling ObjC but disabling OSLog? The whole thing sits inside `if(ENABLE_OBJC)` , that is why I saw no point in a second macro.
Or we want to make these independent?

File: modules/darwinosl/CMakeLists.txt (hunk 5 above)
Comment (HofiOne) [reply modules/darwinosl/CMakeLists.txt:1]: yes, it s the case now, but with these modifications, `ENABLE_OBJC` can depend on multiple future (ObjC) modules

that is why i suggested using a separate `ENABLE_OBJC` variable here

Comment (barracuda156) [reply modules/darwinosl/CMakeLists.txt:1]: Ok, got it. I will deal with this today then.

File: CMakeLists.txt [hunk 6]
```diff
@@ -18,7 +18,19 @@ project(syslog-ng C)
 
 option(ENABLE_CPP "Enable C++" ON)
 
+if (ENABLE_CPP)
+  enable_language(CXX)
+  set(CMAKE_CXX_STANDARD 14 CACHE STRING "C++ standard")
```
Comment (HofiOne) [inline CMakeLists.txt:23]: I do not know why is this different again compared to the autotools version, but can't we create a construct like this here as well?

```
if test "x$enable_cpp" = "xyes"; then
        AX_CXX_COMPILE_STDCXX(11,, mandatory)
        AX_CXX_COMPILE_STDCXX(17,, optional)
elif test "x$enable_cpp" = "xauto"; then
        AX_CXX_COMPILE_STDCXX(11,, optional)
        if test "x$HAVE_CXX11" = "x1"; then
                enable_cpp=yes
        else
                enable_cpp=no
        fi
        AX_CXX_COMPILE_STDCXX(17,, optional)
fi
```

Comment (barracuda156) [reply CMakeLists.txt:23]: There is a way to specify minimum required and desired C++ standard in CMake.

P. S. I think I only added a macro there, not changed C++ standard defaults. Away from the machine atm, difficult to check from mobile.

Comment (HofiOne) [reply CMakeLists.txt:23]: no offense, I know it was not you, just cannot see why there is a difference here

Comment (HofiOne) [reply CMakeLists.txt:23]: as a first-round "fix" could you please just update

set(CMAKE_CXX_STANDARD 14 CACHE STRING "C++ standard")

to

set(CMAKE_CXX_STANDARD 11 CACHE STRING "C++ standard")


after that i will approve and merge the PR

thanks!

**Review by HofiOne — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 6 above)
Comment (barracuda156) [reply CMakeLists.txt:23]: Sure, will do.

(C++11 won’t work for some modules, possibly, since at least Abseil and friends use C++17 now.)

Comment (HofiOne) [reply CMakeLists.txt:23]: never mind, i will do it in one run

**Review by HofiOne — APPROVED** [review]

**Pull request merged**